compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -1686,30 +1710,98 @@ private BType getListConstructorCompatibleNonUnionType(BType type) {
     }
 
     private BType checkArrayType(BLangListConstructorExpr listConstructor, BArrayType arrayType) {
+        int listExprSize = 0;
+        if (arrayType.state != BArrayState.OPEN) {
+            for (BLangExpression expr : listConstructor.exprs) {
+                if (expr.getKind() != NodeKind.LIST_CONSTRUCTOR_SPREAD_OP) {
+                    listExprSize++;
+                    continue;
+                }
+
+                BLangExpression spreadOpExpr = ((BLangListConstructorSpreadOpExpr) expr).expr;
+                BType spreadOpType = checkExpr(spreadOpExpr, this.env);
+                spreadOpType = Types.getReferredType(spreadOpType);
```

Comment:
Do we need this one? we get the referredType in `public BType checkExpr(BLangExpression expr, SymbolEnv env, BType expType, DiagnosticCode diagCode) {`

Reply:
ignore. We are not returning referredType from the above method

Reply:
inside `checkExpr()` we seem to not return the `referredType`.
https://github.com/ballerina-platform/ballerina-lang/blob/709e3d1e6afdf292416f12611ae8e9e1a4a90c0d/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java#L399
is only used in case of an intersection.

Initially, I didn't have the L1723. But when writing the tests I realized referredType is not there.
Refer to the https://github.com/ballerina-platform/ballerina-lang/pull/35601/commits/10347602b60588ec6f1e9905fa4456cc1a604ff1 you can see the affected tests.